### PR TITLE
feat: add config to disable pull queries when validation is required

### LINKED ISSUE #3863: Pull Query: Disable Pull Queries when auth validation is needed on `/query` endpoint
Now that the pull queries are moved to `/query` endpoint, we took a look at what validation happens on this code path. 

We are invoking `KsqlAuthorizationValidator#checkAuthorization()` for every request (push or pull) when either `ksql.access.validator.enable=on` or `ksql.access.validator.enable=auto` and the kafka cluster has a non empty value for `authorizer.class.name`. 

Even as  #3696 avoids eager instantiation of the adminclient (+ other clients) for every request, it simply memoizes it per `ServiceContext` created for every request. When real auth validation is involved, the call path ultimately leads to `KsqlAuthorizationValidatorImpl#checkAccess` , which describes a topic by talking to kafka 

```
 private void checkAccess(
      final ServiceContext serviceContext,
      final String topicName,
      final AclOperation operation
  ) {
    final Set<AclOperation> authorizedOperations = serviceContext.getTopicClient()
        .describeTopic(topicName).authorizedOperations();
    ... 
```

We need to fail pull queries or skip validation, to avoid creating clients per request again.

### PR DESCRIPTION
**Pull request by vinothchandar**
fixes #3863

### Description 
 - Added `ksql.query.pull.skip.access.validator` to control if pull queries work without validation
 - By default, Pull queries error out, if auth validation is needed
 - Replaced DUMMY_VALIDATOR with Optional<> interface for KsqlAuthorizationValidatorFactory
 - Fixed some tests, added test cases

### Testing done 

- [x] Unit tests
- [x] Local testing for performance w and w/o validation enabled 

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by vinothchandar** [comment]
Perf out of box (Kafka does not have ACLs) 

```
15:56:37 [ksql-benchmark]$ wrk -t 8 -c 16 -d 120 --latency -s ./pull-query.lua http://localhost:8088/query
Running 2m test @ http://localhost:8088/query
  8 threads and 16 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     3.68ms    1.95ms  57.60ms   78.50%
    Req/Sec   560.74    159.23     1.37k    54.50%
  Latency Distribution
     50%    3.27ms
     75%    4.47ms
     90%    6.03ms
     99%   10.14ms
  535959 requests in 2.00m, 175.83MB read
Requests/sec:   4463.31
Transfer/sec:      1.46MB
```

**Comment by vinothchandar** [comment]
With validation turned on in Kafka

Pull queries fail 
```
ksql> show tables;

 Table Name       | Kafka Topic      | Format | Windowed 
---------------------------------------------------------
 ORDER_QUANTITIES | ORDER_QUANTITIES | JSON   | false    
---------------------------------------------------------
ksql> SELECT * FROM order_quantities WHERE ROWKEY = 'order-1';
Pull queries are not currently supported, when access validation against Kafka is configured. If you really want to bypass this limitation please set ksql.query.pull.skip.access.validator=true If true, KSQL will  NOT enforce access validation checks for pull queries, which could expose Kafka topics which are secured with RBAC or ACLs. Please enable only after careful consideration. Default value: false
ksql> 
```

and push queries work 

```
ksql> SET 'auto.offset.reset'='earliest';
Successfully changed local property 'auto.offset.reset' from 'earliest' to 'earliest'.
ksql> SELECT * FROM order_quantities EMIT CHANGES LIMIT 10;

+---------------------------------------+---------------------------------------+---------------------------------------+---------------------------------------+
|ROWTIME                                |ROWKEY                                 |ORDER_ID                               |TOTAL_QTY                              |
+---------------------------------------+---------------------------------------+---------------------------------------+---------------------------------------+
|1572465670162                          |order-281.0                            |null                                   |9                                      |
|1572465670437                          |order-615.0                            |null                                   |8                                      |
|1572465677709                          |order-728.0                            |null                                   |7                                      |
|1572465682418                          |order-733.0                            |null                                   |7                                      |
|1572465684522                          |order-364.0                            |null                                   |5                                      |
|1572465698077                          |order-377.0                            |null                                   |0                                      |
|1572465698872                          |order-22.0                             |null                                   |5                                      |
|1572465699660                          |order-5.0                              |null                                   |4                                      |
|1572465701959                          |order-153.0                            |null                                   |5                                      |
|1572465704614                          |order-836.0                            |null                                   |4                                      |
Limit Reached
Query terminated
ksql>
```

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java [hunk 1]
```diff
@@ -257,51 +263,64 @@ public void shouldReturnServiceUnavailableIfTimeoutWaitingForCommandSequenceNumb
...
+        KsqlConfig.KSQL_PULL_QUERIES_SKIP_ACCESS_VALIDATOR_CONFIG, true
+    )));
 
-    // Then:
+    // Expect
     verify(serviceContext, never()).getAdminClient();
-  }
+    verify(serviceContext, never()).getConnectClient();
+    verify(serviceContext, never()).getSchemaRegistryClient();
+    verify(serviceContext, never()).getTopicClient();
```
Comment (big-andy-coates) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java]: test bug: The `verify` calls need to come _after_ the test case.  This is verifying the calls are not being made by the test setup.

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java [hunk 2]
```diff
@@ -257,51 +263,64 @@ public void shouldReturnServiceUnavailableIfTimeoutWaitingForCommandSequenceNumb
...
-    testResource.streamQuery(
+    final Response response = testResource.streamQuery(
         serviceContext,
         new KsqlRequest(PULL_QUERY_STRING, Collections.emptyMap(), null)
     );
 
     // Then:
-    verify(serviceContext, never()).getSchemaRegistryClient();
+    assertThat(response.getStatus(), is(Errors.badRequest("").getStatus()));
+    final KsqlErrorMessage expectedEntity = (KsqlErrorMessage) response.getEntity();
```
Comment (big-andy-coates) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java]: consider adding `assertThat(response.getEntity(), is(instanceOf(KsqlErrorMessage.class)));` before this line to get a better error message than a class cast exception

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java [hunk 3]
```diff
@@ -257,51 +263,64 @@ public void shouldReturnServiceUnavailableIfTimeoutWaitingForCommandSequenceNumb
...
     // When:
-    testResource.streamQuery(
+    final Response response = testResource.streamQuery(
         serviceContext,
         new KsqlRequest(PULL_QUERY_STRING, Collections.emptyMap(), null)
     );
 
     // Then:
-    verify(serviceContext, never()).getTopicClient();
+    assertThat(response.getStatus(), is(Errors.badRequest("").getStatus()));
```
Comment (big-andy-coates) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java]: Would be nice if this test tested the positive path, i.e. that pull queries _work_ if not validating...

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java [hunk 4]
```diff
@@ -133,14 +135,18 @@
   @Mock
   private KsqlAuthorizationValidator authorizationValidator;
   private StreamedQueryResource testResource;
-  private PreparedStatement<Statement> statement;
+  private PreparedStatement<Statement> pushQueryStatement;
```
Comment (big-andy-coates) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java]: Can we revert this change please? It's confusing to have `pushQueryStatement` used when testing PRINT TOPIC!

```
pushQueryStatement = PreparedStatement.of("print", cmd);
```

The test doesn't need specific fields for specific statement types.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java [hunk 5]
```diff
@@ -213,10 +213,10 @@ public void onOpen(final Session session, final EndpointConfig unused) {
 
       serviceContext = createServiceContext(session.getUserPrincipal());
 
-      authorizationValidator.checkAuthorization(
+      authorizationValidator.ifPresent(validator -> validator.checkAuthorization(
```
Comment (big-andy-coates) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java]: This websocket endpoint handles pull queries too, so should have consistent handling of auth checks as the rest endpoint.

File: ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java [hunk 6]
```diff
@@ -198,6 +198,13 @@
           + "\"off\" disables the validator. If set to \"auto\", KSQL will attempt to discover "
           + "whether the Kafka cluster supports the required API, and enables the validator if "
           + "it does.";
+  public static final String KSQL_PULL_QUERIES_SKIP_ACCESS_VALIDATOR_CONFIG =
+      "ksql.query.pull.skip.access.validator";
+  public static final boolean KSQL_PULL_QUERIES_SKIP_ACCESS_VALIDATOR_DEFAULT = false;
+  public static final String KSQL_PULL_QUERIES_SKIP_ACCESS_VALIDATOR_DOC = "If true, KSQL will "
+      + " NOT enforce access validation checks for pull queries, which could expose Kafka topics"
```
Comment (big-andy-coates) [inline ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java]: `validation checks` here is a bit vague.  Can we be more explicit?  This is really about skips `authorization checks`.  The docs and the name of this config should reflect this.

I think the actual functionality is that with this set to the default `false` KSQL won't support pull queries when running against a secure Kafka. With it set to `true` KSQL won't check the user has access to the topics underlying the the materialized state the pull query is accessing.

Is that right?

**Comment by apurvam** [comment]
> It looks like this PR disables pull queries for any KSQL server running against a secured Kafka cluster, e.g. one secured with ACLs or some custom auth mech, i.e. it disabled pull queries for anyone using KSQL in a modern production environment. Is this really what we want? I guess we're releasing this as a preview and we can look to quickly iterate and remove this restriction. But just wanted to call this out and make sure this is indeed our intent with this PR.

This is accurate. The main worry is if someone runs a pull query benchmark against a production cluster, the current mechanism could actually destabilize the cluster due to its design. IMO the reputations damage due to that could be worse that having the feature be opt in with the risks understood. 

@derekjn is on board with this. cc @MichaelDrogalis for visibility in case he missed the other threads .

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java (hunk 2 above)
Comment (vinothchandar) [reply ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java]: done.

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java (hunk 3 above)
Comment (vinothchandar) [reply ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java]: So, this case covers that we will get past the config check I introduced.. Ultimately, this test itself mocks a whole bunch of things that you need for pull queries to actually succeed. This in combination with other tests that test pull queries functionally are sufficient IMO

**Comment by derekjn** [comment]
LGTM 👍

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java (hunk 5 above)
Comment (apurvam) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java]: Let’s keep parity for now.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java [hunk 7]
```diff
@@ -175,23 +177,35 @@ private Response handleStatement(
...
       if (statement.getStatement() instanceof Query) {
         final PreparedStatement<Query> queryStmt = (PreparedStatement<Query>) statement;
 
         if (queryStmt.getStatement().isPullQuery()) {
+          final boolean skipAccessValidation = ksqlConfig.getBoolean(
+              KsqlConfig.KSQL_PULL_QUERIES_SKIP_ACCESS_VALIDATOR_CONFIG);
+          if (authorizationValidator.isPresent() && !skipAccessValidation) {
+            return Errors.badRequest("Pull queries are not currently supported when "
+                + "access validation against Kafka is configured. If you really want to "
+                + "bypass this limitation please set "
```
Comment (spena) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java]: Can you add a warning message or specify the security reasons for disabling the check here?. Users won't see the config doc when seeing this error. They might just go and change this config without knowing the security reasons behind this.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java (hunk 5 above)
Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java]: This endpoint is used by the KSQL UI. Better to support it now than forget it later.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java (hunk 7 above)
Comment (vinothchandar) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java]: We already print the doc of the class also, which calls out what this exposes.

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java (hunk 3 above)
Comment (big-andy-coates) [reply ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResourceTest.java]: OK
